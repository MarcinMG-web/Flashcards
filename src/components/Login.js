import React, { useState, useRef } from 'react';
import { Link, Redirect } from 'react-router-dom';

import { getToken, isUser } from '../../src/services/Authorization';

const Login = () => {
  const [loginParams, setLoginParams] = useState({
    username: '',
    password: '',
  });

  const [errorValid, setErrorValid] = useState('');

  const [isLogIn, setIsLogIn] = useState(false);

  const [spinier, setSpinier] = useState(false);

  const formRef = useRef();

  const handleChange = (e) => {
    setLoginParams({
      ...loginParams,
      [e.target.name]: e.target.value,
    });
  };

  const isLogInToApplication = (e) => {
    e.preventDefault();

    const getDataToken = async () => {
      /**
       * Login
       */

      setSpinier(true);
      const isUserResponse = await isUser(loginParams.username);
      const isValidUser = isUserResponse['data'][loginParams.username];
      console.log(isValidUser);
      await getToken(loginParams);

      const savedTokenFromLocalStorage = localStorage.getItem('token');
      if (savedTokenFromLocalStorage !== null) {
        setIsLogIn(true);
        setSpinier(false);
      } else {
        setIsLogIn(false);
        setErrorValid('Try again to get access or create account :)');
        setSpinier(false);
      }
    };

    getDataToken();

    if (formRef.current) {
      formRef.current.reset();
    }
  };

  if (isLogIn) {
    return <Redirect from='/' to='/Dashboard' />;
  }

  return (
    <div className='ContainerLogin'>
      <div className='ContainerLogin_header'>Login Panel:</div>

      <div className='ContainerLogin_main'>
        <div className='ContainerLogin_main-form'>
          <form onSubmit={isLogInToApplication} ref={formRef}>
            <div className='ContainerLogin_main-form-input'>
              <label style={{ margin: '15px' }}>Enter your user name:</label>
              <input
                type='text'
                placeholder='username'
                name='username'
                onChange={handleChange}
                className='inputLogin'
              />
            </div>
            <div className='ContainerLogin_main-form-input'>
              <label style={{ margin: '15px' }}>Enter your password:</label>
              <input
                type='password'
                placeholder='password'
                name='password'
                onChange={handleChange}
                className='inputLogin'
              />

              <div
                className='fa-3x'
                style={{ color: 'orange', fontSize: '25px' }}
              >
                {spinier ? <i className='fa fa-spinner fa-pulse'></i> : ''}
              </div>

              <span style={{ color: 'red', fontSize: '14px' }}>
                {errorValid}
              </span>

              <div className='ContainerLogin_main-form-button'>
                <Link
                  to='./Registration'
                  type='submit'
                  className='btn_createAccount'
                >
                  Create Account
                </Link>
                <button type='submit' className='btn_start'>
                  Let's get started!
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
      <div className='ContainerLogin_footer'></div>
    </div>
  );
};

export default Login;
