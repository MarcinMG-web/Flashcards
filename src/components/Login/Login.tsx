import { AxiosResponse } from 'axios';
import React, { useState, useEffect, useRef } from 'react';
import { Link, Redirect } from 'react-router-dom';

import { getToken, isUser } from '../../services/Authorization';

const Login: React.FC = () => {
  const [loginParams, setLoginParams] = useState({
    username: '',
    password: '',
  });

  const [errorValid, setErrorValid] = useState('');

  const [isLogIn, setIsLogIn] = useState(false);

  const [spinier, setSpinier] = useState(false);

  const formRef = useRef<HTMLInputElement | any>();

  const [didMount, setDidMount] = useState(false);

  useEffect(() => {
    setDidMount(true);

    setTimeout(() => {
      setSpinier(false);
    }, 300);

    return () => setDidMount(false);
  }, [spinier]);

  /**
   * Set up login params
   */
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setLoginParams({
      ...loginParams,
      [e.target.name]: e.target.value,
    });
  };

  /**
   * Login
   */
  const getDataToken = async (): Promise<void> => {
    setSpinier(true);
    const isUserResponse: AxiosResponse | any = await isUser(
      loginParams.username
    );
    const isValidUser: boolean = isUserResponse['data'][loginParams.username];
    await getToken(loginParams);

    const savedTokenFromLocalStorage: string | null =
      localStorage.getItem('token');
    if (savedTokenFromLocalStorage !== null) {
      setIsLogIn(true);
      setErrorValid('Try again set the credentials');
      setSpinier(false);
    } else if (isValidUser) {
      setIsLogIn(false);
      setErrorValid('Incorrect password. Try again');
      setSpinier(false);
    } else {
      setIsLogIn(false);
      setErrorValid('User does not exist. Try to register first');
      setSpinier(false);
    }
  };

  const isLogInToApplication = (e: React.SyntheticEvent<EventTarget>): void => {
    e.preventDefault();

    getDataToken();

    if (formRef.current) {
      formRef.current.reset();
    }
  };

  if (isLogIn) {
    return <Redirect from='/' to='/Dashboard' />;
  }

  if (!didMount) {
    return null;
  }

  return (
    <div className='ContainerLogin'>
      <div className='ContainerLogin_header'>Login Panel</div>

      <div className='ContainerLogin_main'>
        <div className='ContainerLogin_main-form'>
          <form onSubmit={isLogInToApplication} ref={formRef}>
            <div className='ContainerLogin_main-form-input'>
              <label style={{ margin: '15px' }}>Enter your user name:</label>
              <input
                type='text'
                placeholder='username'
                name='username'
                onChange={handleChange}
                className='inputLogin'
                data-testid='input-username'
              />
            </div>
            <div className='ContainerLogin_main-form-input'>
              <label style={{ margin: '15px' }}>Enter your password:</label>
              <input
                type='password'
                placeholder='password'
                name='password'
                onChange={handleChange}
                className='inputLogin'
                data-testid='input-password'
              />

              <div
                className='fa-3x'
                style={{ color: 'orange', fontSize: '25px' }}
              >
                {spinier ? <i className='fa fa-spinner fa-pulse'></i> : ''}
              </div>

              <span style={{ color: 'red', fontSize: '14px' }} data-testid='span-validation-message'>
                {errorValid}
              </span>

              <div className='ContainerLogin_main-form-button'>
                <Link
                  to='./Registration'
                  type='submit'
                  className='btn_createAccount'
                >
                  Create Account
                </Link>
                <button type='submit' className='btn_start'>
                  Let's get started!
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
      <div className='ContainerLogin_footer'></div>
    </div>
  );
};

export default Login;
