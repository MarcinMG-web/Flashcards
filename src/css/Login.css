.ContainerLogin {
  width: 600px;
  height: 100vh;
  min-width: 300px;
  background-color: #fff;
  color: #000;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
}

.ContainerLogin_header {
  width: 100%;
  height: 60px;
  background-color: #000;
  color: #fff;
  display: flex;
  justify-content: center;
  align-items: center;
}

.ContainerLogin_main {
  display: flex;
  flex-direction: column;
}

.ContainerLogin_main-form {
  width: 35vw;
}

.ContainerLogin_main-form-button {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  margin-top: 10px;
}

.ContainerLogin_footer {
  width: 100%;
  height: 60px;
  background-color: black;
  display: flex;
  justify-content: center;
  align-items: center;
}

.inputLogin {
  width: 100%;
  height: 40px;
  border: 1px solid orange;
  border-radius: 5px;
  text-align: center;
  padding: 10px;
  margin-top: 10px;
  margin-bottom: 10px;
  font-family: 'New Tegomin', serif;
  font-size: 16px;
}

.special {
  display: none;
}
