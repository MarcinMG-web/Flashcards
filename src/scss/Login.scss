.ContainerLogin {
  width: 600px;
  height: 100vh;
  min-width: 300px;
  background-color: #fff;
  color: #000;

  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;

  &_header {
    width: 100%;
    height: 60px;
    background-color: #000;
    color: #fff;
    display: flex;
    justify-content: center;
    align-items: center;
  }
  &_main {
    display: flex;
    flex-direction: column;
    &-form {
      width: 35vw;
      &-button {
        display: flex;
        flex-direction: row;
        align-items: center;
        justify-content: center;
        margin-top: 10px;
      }
    }
  }

  &_footer {
    width: 100%;
    height: 60px;
    background-color: #fff;
    display: flex;
    justify-content: center;
    align-items: center;
  }
}

.inputLogin {
  width: 100%;
  height: 40px;
  border: 1px solid rgb(255, 165, 0);
  border-radius: 5px;

  text-align: center;
  padding: 10px;
  margin-top: 10px;
  margin-bottom: 10px;
  font-family: 'New Tegomin', serif;
  font-size: 16px;
}

.special {
  display: none;
}

$phone-width: 466px;
$phone-width-1: 465px;
$tablet-width: 768px;
$desktop-width: 1024px;

@media (max-width: $phone-width) {
  .ContainerLogin {
    width: 100vw;
    height: 100vh;
    margin: 0;

    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-between;

    &_header {
      width: 100%;
      height: 60px;
      background-color: #000;
      color: #fff;
      display: flex;
      justify-content: center;
      align-items: center;
    }
    &_main {
      display: flex;
      flex-direction: column;
      &-form {
        width: 90vw;
        &-button {
          display: flex;
          flex-direction: row;
          align-items: center;
          justify-content: center;
          margin-top: 10px;
        }
      }
    }

    &_footer {
      width: 100%;
      height: 60px;
      background-color: #fff;
      display: flex;
      justify-content: center;
      align-items: center;
    }
  }
}

@media (min-width: $phone-width-1) and (max-width: $tablet-width) {
  .ContainerLogin {
    width: 100vw;
    height: 100vh;
    margin: 0;

    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-between;

    &_header {
      width: 100%;
      height: 60px;
      background-color: #000;
      color: #fff;
      display: flex;
      justify-content: center;
      align-items: center;
    }
    &_main {
      display: flex;
      flex-direction: column;
      &-form {
        width: 90vw;
        &-button {
          display: flex;
          flex-direction: row;
          align-items: center;
          justify-content: center;
          margin-top: 10px;
        }
      }
    }

    &_footer {
      width: 100%;
      height: 60px;
      background-color: #fff;
      display: flex;
      justify-content: center;
      align-items: center;
    }
  }
}
