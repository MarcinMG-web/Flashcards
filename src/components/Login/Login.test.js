import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import Login from './Login';

import { BrowserRouter } from 'react-router-dom';

const MockLogin = () => {
  return (
    <BrowserRouter>
      <Login />
    </BrowserRouter>
  );
};

describe('Login Tests - render', () => {
  it('snapshot', () => {
    const tree = render(<MockLogin />);

    expect(tree).toMatchSnapshot();
  });

  it('render without crashing', () => {
    render(<MockLogin />);
  });
});

describe('Tests Elements UI', () => {
  it('render input username', () => {
    render(<MockLogin />);

    const checkInputName = screen.getByTestId('input-username');

    expect(checkInputName).toBeInTheDocument();
  });

  it('render input password', () => {
    render(<MockLogin />);

    const checkInputPassword = screen.getByTestId('input-password');

    expect(checkInputPassword).toBeInTheDocument();
  });

  it('render button Create account', () => {
    render(<MockLogin />);

    const buttonCreateAccount = screen.getByText(/Create Account/i);

    expect(buttonCreateAccount).toBeInTheDocument();
  });

  it('render button Lets get started', () => {
    render(<MockLogin />);

    const buttonLetsGetStarted = screen.getByText(/Let's get started!/i);

    expect(buttonLetsGetStarted).toBeInTheDocument();
  });
});

describe('Tests Business Logic', () => {
  it('render input element with value', () => {
    render(<MockLogin />);

    const checkInputName = screen.getByTestId('input-username');
    fireEvent.change(checkInputName, { target: { value: 'Admin' } });

    expect(checkInputName.value).toBe('Admin');
  });

  it('render input with value after clicked', () => {
    render(<MockLogin />);

    const checkInputName = screen.getByTestId('input-username');
    const buttonLetsGetStarted = screen.getByText(/Let's get started!/i);

    fireEvent.change(checkInputName, { target: { value: 'Admin' } });
    fireEvent.click(buttonLetsGetStarted);

    expect(checkInputName.value).toBe('Admin');
  });

  it('render validation info when user does not exist', async () => {
    render(<MockLogin />);

    const checkInputName = screen.getByTestId('input-username');
    const buttonLetsGetStarted = screen.getByText(/Let's get started!/i);
    const validationInfo = 'User does not exist. Try to register first';

    fireEvent.change(checkInputName, { target: { value: 'Random' } });
    fireEvent.click(buttonLetsGetStarted);
    

    await waitFor(() => {
      const spanMassageValidation = screen.getByTestId(
        'span-validation-message'
      );

      expect(spanMassageValidation).toHaveTextContent(validationInfo);
    });
  });

  it('render validation info when you forget add password or password is incorrect', async () => {
    render(<MockLogin />);

    const checkInputName = screen.getByTestId('input-username');
    const buttonLetsGetStarted = screen.getByText(/Let's get started!/i);
    const validationInfo = 'Incorrect password. Try again';

    fireEvent.change(checkInputName, { target: { value: 'Admin' } });
    fireEvent.click(buttonLetsGetStarted);

    await waitFor(() => {
      const spanMassageValidation = screen.getByTestId(
        'span-validation-message'
      );

      expect(spanMassageValidation).toHaveTextContent(validationInfo);
    });
  });
});
