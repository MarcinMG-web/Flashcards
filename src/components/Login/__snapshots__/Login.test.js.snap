// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Tests - render snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ContainerLogin"
      >
        <div
          class="ContainerLogin_header"
        >
          Login Panel
        </div>
        <div
          class="ContainerLogin_main"
        >
          <div
            class="ContainerLogin_main-form"
          >
            <form>
              <div
                class="ContainerLogin_main-form-input"
              >
                <label
                  style="margin: 15px;"
                >
                  Enter your user name:
                </label>
                <input
                  class="inputLogin"
                  data-testid="input-username"
                  name="username"
                  placeholder="username"
                  type="text"
                />
              </div>
              <div
                class="ContainerLogin_main-form-input"
              >
                <label
                  style="margin: 15px;"
                >
                  Enter your password:
                </label>
                <input
                  class="inputLogin"
                  data-testid="input-password"
                  name="password"
                  placeholder="password"
                  type="password"
                />
                <div
                  class="fa-3x"
                  style="color: orange; font-size: 25px;"
                />
                <span
                  data-testid="span-validation"
                  style="color: red; font-size: 14px;"
                />
                <div
                  class="ContainerLogin_main-form-button"
                >
                  <a
                    class="btn_createAccount"
                    href="/Registration"
                    type="submit"
                  >
                    Create Account
                  </a>
                  <button
                    class="btn_start"
                    type="submit"
                  >
                    Let's get started!
                  </button>
                </div>
              </div>
            </form>
          </div>
        </div>
        <div
          class="ContainerLogin_footer"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ContainerLogin"
    >
      <div
        class="ContainerLogin_header"
      >
        Login Panel
      </div>
      <div
        class="ContainerLogin_main"
      >
        <div
          class="ContainerLogin_main-form"
        >
          <form>
            <div
              class="ContainerLogin_main-form-input"
            >
              <label
                style="margin: 15px;"
              >
                Enter your user name:
              </label>
              <input
                class="inputLogin"
                data-testid="input-username"
                name="username"
                placeholder="username"
                type="text"
              />
            </div>
            <div
              class="ContainerLogin_main-form-input"
            >
              <label
                style="margin: 15px;"
              >
                Enter your password:
              </label>
              <input
                class="inputLogin"
                data-testid="input-password"
                name="password"
                placeholder="password"
                type="password"
              />
              <div
                class="fa-3x"
                style="color: orange; font-size: 25px;"
              />
              <span
                data-testid="span-validation"
                style="color: red; font-size: 14px;"
              />
              <div
                class="ContainerLogin_main-form-button"
              >
                <a
                  class="btn_createAccount"
                  href="/Registration"
                  type="submit"
                >
                  Create Account
                </a>
                <button
                  class="btn_start"
                  type="submit"
                >
                  Let's get started!
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
      <div
        class="ContainerLogin_footer"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
